docs: add Keycloak configuration guide (#532)
rhel-lightspeed/linux-mcp-server: docs/keycloak.md

# Keycloak Configuration

> This guide covers how to configure Keycloak for JWT  and Introspection based authentication.

Reviewer: Needs expansion and context: 
```suggestion
[Keycloak](https://www.keycloak.org) is a open-source identity and access management solution that can be used as the OAuth2 authorization server for linux-mcp-server. This section describes how to configure a standalone instance of KeyCloak for testing purposes. It will also be useful if your organization is already using KeyCloak and you want to add linux-mcp-server to it.
```

## Prerequisites

> - Keycloak server runnings

Reviewer: ```suggestion
- Keycloak server running
```

- Admin access to Keycloak Admin Console

### Running Keycloak Locally with Podman

```bash
# Pull the latest Keycloak image
podman pull quay.io/keycloak/keycloak:latest

# Run Keycloak in development mode
podman run -d \
  --name keycloak \
  -p 8080:8080 \
  -e KEYCLOAK_ADMIN=admin \
  -e KEYCLOAK_ADMIN_PASSWORD=admin \
  quay.io/keycloak/keycloak:latest start-dev

# Check logs
podman logs -f keycloak

# Access admin console at: http://localhost:8080/admin
# Login with: admin / admin
```

---

This approach uses **Dynamic Client Registration (DCR)** where MCP clients automatically register themselves with Keycloak. The MCP server validates JWT tokens or uses introspection.

### Step 1: Create Realm

> 1. Navigate to **Keycloak Admin Console**: `http://localhost:8080/admin/master/console/`

Reviewer: I think you should have a sentence here to say what a realm is.

2. Log in with admin credentials
3. Click the realm dropdown (top left, shows "master") → **Create Realm**
4. Configure:

>    - **Realm name**: Choose your realm name

Reviewer: Provide an example

### Step 4: Create a Test User

Create a user with email and password for testing authentication:

- Navigate to **Users** and create a new user
- Set a non-temporary password in the **Credentials** tab

## MCP Server Configuration

###  JWT Validation

The MCP server validates JWT tokens using Keycloak's public keys (JWKS).

```bash
# Transport
export LINUX_MCP_TRANSPORT=streamable-http
export LINUX_MCP_HOST=localhost
export LINUX_MCP_PORT=3000

# JWT Authentication
export LINUX_MCP_AUTH__PROVIDER=jwt
export LINUX_MCP_AUTH__JWT__JWKS_URI="http://localhost:8080/realms/mcp_realm/protocol/openid-connect/certs"
export LINUX_MCP_AUTH__JWT__ISSUER="http://localhost:8080/realms/mcp_realm"
export LINUX_MCP_AUTH__JWT__AUDIENCE="account"  # Must match the audience mapper value

```

**Authentication Flow**: Clients discover Keycloak via `/.well-known/oauth-protected-resource/mcp`, auto-register using DCR, obtain JWT tokens through Keycloak's OAuth flow, and send tokens to the MCP server which validates them locally using JWKS.

### Token Introspection

> The MCP server validates tokens by calling Keycloak's introspection endpoint. Requires a confidential client.

Reviewer: Should briefly say what a confidential client is

# Keycloak Configuration

[Keycloak](https://www.keycloak.org) is a open-source identity and access management solution that can be used as the OAuth2 authorization server for linux-mcp-server. This section describes how to configure a standalone instance of KeyCloak for testing purposes. It will also be useful if your organization is already using KeyCloak and you want to add linux-mcp-server to it.

## Prerequisites

- Keycloak server running
- Admin access to Keycloak Admin Console

### Running Keycloak Locally with Podman

```bash
# Pull the latest Keycloak image
podman pull quay.io/keycloak/keycloak:latest

# Run Keycloak in development mode
podman run -d \
  --name keycloak \
  -p 8080:8080 \
  -e KEYCLOAK_ADMIN=admin \
  -e KEYCLOAK_ADMIN_PASSWORD=admin \
  quay.io/keycloak/keycloak:latest start-dev

# Check logs
podman logs -f keycloak

# Access admin console at: http://localhost:8080/admin
# Login with: admin / admin
```

---

> This approach uses **Dynamic Client Registration (DCR)** where MCP clients automatically register themselves with Keycloak. The MCP server validates JWT tokens or uses introspection.

Reviewer: This paragraph mixes two things, both of which could do with some expansion.

```suggestion
The approach we'll take here is using **Dynamic Client Registration (DCR)**. With DCR, MCP clients automatically register themselves with the authorization. Otherwise, clients would have to be added through the admin console, which is not typically supported for chat clients like Goose or Claude Desktop.

Either the jwt or introspection authorization backend for linux-mcp-server can be used with KeyCloak. The difference is slight - the jwt backend is easier to set up and might offer a bit better performance, while the introspection backend offers immediate token revocation without waiting for token expiration if, for example, a user account is removed.
```

Reviewer: Meant to add this comment on the last review, but it was still sitting pending.

### Step 1: Create Realm

> A realm is Keycloak's top-level unit of tenancy, it manages its own set of users, clients, roles, and authentication settings, isolated from other realms on the same server.

Reviewer: "unit of tenancy" is obscure - here's a suggestion from Gemini - "A realm is an isolated workspace inside Keycloak. It acts as a standalone container with its own users, apps, and login rules that don't mix with anything else on the server."
